File: leetcode/70-climbing-stairs.ts
/**
 * 爬楼梯
 * https://leetcode.cn/problems/climbing-stairs/
 */

// 空间复杂父为O(1)
export function climbStairs(n: number): number {
  let p = 0,
    q = 0,
    r = 1
  for (let i = 1; i <= n; i++) {
    p = q
    q = r
    r = p + q
  }
  return r
}

console.log(climbStairs(4))

File: leetcode-hot-100/70-climbing-stairs.py
# 70. 爬楼梯
# https://leetcode.cn/problems/climbing-stairs


class Solution:
    def climbStairs(self, n: int) -> int:
        if n == 1:
            return 1
        if n == 2:
            return 2
        dp = [1] * (n + 1)
        dp[1], dp[2] = 1, 2
        print(dp)
        for i in range(3, n + 1):
            dp[i] = dp[i - 1] + dp[i - 2]
        return dp[n]


print(Solution().climbStairs(3))
